Use pattern rules in Makefiles to reduce specific rule

NiKiZe opened the pull request:
Minimizes number of lines, and reduces maintenance for new protocols
This should fix #1109 

There is more work that could be done to improve things.
One observation is that RC5/RC6 is specified for two protocols in tools, but only one in tests?

This might also have added generic dependencies which wasn't there before,
but that should be a small price to pay, and should be fixable in the future.

NiKiZe committed:
Use pattern rules in Makefiles to reduce specific rule

NiKiZe committed:
Automate Makefiles by using wildcards to find targets

crankyoldgit commented:
The RC5 thing may have predated me including all the ir_*.o and ir_*.h files as dependencies. Feel free to include (remove) them in the cleanup of it works with your new method. Which btw is all colours of AWESOME so far!

NiKiZe commented:
For now I don't think that we need to go much deeper into the RC5 stuff.
Instead I think that should be improved by automate more to find correct dependencies for each file.

crankyoldgit commented:
Sounds good to me.

crankyoldgit reviewed (approved):
I'll check this out properly before I merge it in the morning when I'm awake, but looks good to me so far. Many many thanks for this!

crankyoldgit committed:
More improvements to Makefiles.
* Keep all intermediate files. e.g. .o files.
  - GNU make seems to delete any intermediate file not explicitly named
    in the Makefile. This causes extra recompilation when in some circumstances.
* Remove single-quote marks from around `$*` as it shows up weird in the output.
  - This is totally cosmetic.
* Remove the special-case ir_Protocol cases as they are not that special.

crankyoldgit commented:
@NiKiZe I've made some subtle improvements to your PR.
I now know about 5x more about make & Makefiles than I did 24 hours ago. I'm not sure that's a good thing. ;-) But thanks!
That said. 5x a very small amount is still a small amount. :-P

NiKiZe commented:
I keept the quotes to make it easy to distinguish between automatic and specific rules. (But my fault for not documenting that)


Thanks!

crankyoldgit commented:
FYI, the changes mentioned above have now been included in the latest release (v2.7.7) of the library.
